zio-config/src/test/scala/zio/config/NestedConfigTest.scala

```diff
@@ -0,0 +1,145 @@
+      port <- Gen.anyInt
+    } yield DbConnection(host, port)
+
+  val genDb: Gen[Random, Database] =
+    for {
+      connection  <- Gen.either(genNonEmptyString(20).map(DbUrl), genDbConnection)
+      credentials <- Gen.option(genCredentials)
+    } yield Database(connection, credentials)
+
+  val genAppConfig: Gen[Random, AppConfig] =
+    for {
+      db      <- genDb
+      pricing <- Gen.anyFloat.map(_.toDouble)
+    } yield AppConfig(db, pricing)
+
+  final case class TestParams(keys: KeyParams, value: AppConfig) {
+
+    def config: ConfigDescriptor[String, String, AppConfig] = {
+      val credentials  = (string(keys.user) |@| string(keys.password))(Credentials.apply, Credentials.unapply)
+      val dbConnection = (string(keys.host) |@| int(keys.port))(DbConnection.apply, DbConnection.unapply)
+
+      val database =
+        (string(keys.connection).xmap(DbUrl)(_.value).orElseEither(nested(keys.connection)(dbConnection)) |@|
+          nested(keys.credentials)(credentials).optional)(Database.apply, Database.unapply)
+
+      (nested(keys.database)(database) |@| double(keys.pricing))(AppConfig, AppConfig.unapply)
+    }
+
+    def record: Record[String, String] = {
+      val connection =
+        keys.connection -> value.db.connection.fold(
+          url => Leaf(url.value),
+          dbConnection =>
+            Record(Map(keys.host -> Leaf(dbConnection.host), keys.port -> Leaf(dbConnection.port.toString)))
+        )
+
+      val credentials = value.db.credentials.map { v =>
+        keys.credentials -> Record(Map(keys.user -> Leaf(v.user), keys.password -> Leaf(v.password)))
+      }
+
+      Record(
+        Map(
+          keys.database -> Record(Seq(Option(connection), credentials).flatten.toMap),
+          keys.pricing  -> Leaf(value.pricing.toString)
+        )
+      )
+    }
+
+    def source: ConfigSource[String, String] =
+      ConfigSource.fromMap(
+        Seq(
+          value.db.connection.fold(
+            url => Seq(s"${keys.database}.${keys.connection}" -> url.value),
+            connection => {
+              val connectionKey = s"${keys.database}.${keys.connection}"
+              Seq(
+                s"$connectionKey.${keys.host}" -> connection.host,
+                s"$connectionKey.${keys.port}" -> connection.port.toString
+              )
+            }
+          ),
+          value.db.credentials.fold(Seq.empty[(String, String)]) { c =>
+            val credentialsKey = s"${keys.database}.${keys.credentials}"
+            Seq(
+              s"$credentialsKey.${keys.user}"     -> c.user,
+              s"$credentialsKey.${keys.password}" -> c.password
+            )
+          },
+          Seq(keys.pricing -> value.pricing.toString)
+        ).flatten.toMap
+      )
+  }
+
+  private val genKey = genSymbol(1, 20)
+
+  val genConfigKeys: Gen[Random, KeyParams] =
+    for {
+      kHost        <- genKey
+      kPort        <- genKey.filter(_ != kHost)
+      kUser        <- genKey
```

Comment:
`kHost` and `kUser` are not on the same "level" in configuration so they should have the same value?
Also, I was thinking that in this case key's generation may not have any benefit, but it can make code more complicated. Maybe we should have fixed the key's values.